src/detect-bytemath.c

```diff
@@ -0,0 +1,1284 @@
+        DetectPcreData *pd = (DetectPcreData *)prev_pm->ctx;
+        pd->flags |= DETECT_PCRE_RELATIVE_NEXT;
+    }
+
+ okay:
+    return 0;
+
+ error:
+    if (rvalue)
+        SCFree(rvalue);
+    DetectByteMathFree(data);
+    return ret;
+}
+
+/**
+ * \brief Used to free instances of DetectByteMathractData.
+ *
+ * \param ptr Instance of DetectByteMathData to be freed.
+ */
+static void DetectByteMathFree(void *ptr)
+{
+    if (ptr != NULL) {
+        DetectByteMathData *bmd = ptr;
+        if (bmd->result != NULL)
+            SCFree((void *)bmd->result);
+        SCFree(bmd);
+    }
+
+    return;
+}
+
+/**
+ * \brief Lookup the SigMatch for a named byte_math variable.
+ *
+ * \param arg The name of the byte_math variable to lookup.
+ * \param s Pointer the signature to look in.
+ *
+ * \retval A pointer to the SigMatch if found, otherwise NULL.
+ */
+SigMatch *DetectByteMathRetrieveSMVar(const char *arg, const Signature *s)
+{
+    const int nlists = s->init_data->smlists_array_size;
+    for (int list = 0; list < nlists; list++) {
+        SigMatch *sm = s->init_data->smlists[list];
+        while (sm != NULL) {
+            if (sm->type == DETECT_BYTEMATH) {
+                const DetectByteMathData *bmd = (const DetectByteMathData *)sm->ctx;
+                if (strcmp(bmd->result, arg) == 0) {
+                    SCLogDebug("Retrieved SM for \"%s\"", arg);
+                    return sm;
+                }
+            }
+            sm = sm->next;
+        }
+    }
+
+    return NULL;
+}
+
+/*************************************Unittests********************************/
+#ifdef UNITTESTS
+
+#include "app-layer-parser.h"
+
+static int DetectByteMathParseTest01(void)
+{
+
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 4, offset 2, oper +,"
+                                                  "rvalue 10, result bar", NULL);
+    FAIL_IF(bmd == NULL);
+
+    FAIL_IF_NOT(bmd->nbytes == 4);
+    FAIL_IF_NOT(bmd->offset == 2);
+    FAIL_IF_NOT(bmd->oper == DETECT_BYTEMATH_OPERATOR_PLUS);
+    FAIL_IF_NOT(bmd->rvalue == 10);
+    FAIL_IF_NOT(strcmp(bmd->result, "bar") == 0);
+    FAIL_IF_NOT(bmd->endian == DETECT_BYTEMATH_ENDIAN_DEFAULT);
+    FAIL_IF_NOT(bmd->base == DETECT_BYTEMATH_BASE_DEFAULT);
+
+    DetectByteMathFree(bmd);
+
+    PASS;
+}
+
+static int DetectByteMathParseTest02(void)
+{
+    /* bytes value invalid */
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 257, offset 2, oper +, "
+                                                  "rvalue 39, result bar", NULL);
+
+    if (bmd != NULL)
+        DetectByteMathFree(bmd);
+
+    FAIL_IF(bmd != NULL);
+    PASS;
+}
+
+static int DetectByteMathParseTest03(void)
+{
+    /* bytes value invalid */
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 11, offset 2, oper +, "
+                                                  "rvalue 39, result bar", NULL);
+    if (bmd != NULL)
+        DetectByteMathFree(bmd);
+
+    FAIL_IF(bmd != NULL);
```

Comment:
this should move up, the the `if (bmd != NULL)` condition can be removed.

General rule in (new) unittests: no conditions.

Reply:
Also, tests should clean up memory on their success path. But can ignore failure path.